ceed-fortran.c

```diff
@@ -0,0 +1,523 @@
+    CeedRealloc(CeedBasis_count_max, &CeedBasis_dict);
+  }
+
+  *err = CeedBasisCreateTensorH1(Ceed_dict[*ceed], *dim, *ndof, *P1d, *Q1d, interp1d, grad1d,
+             qref1d, qweight1d, &CeedBasis_dict[CeedBasis_count]);
+
+  if (*err == 0) {
+    *basis = CeedBasis_count++;
+    CeedBasis_n++;
+  }
+}
+
+#define fCeedBasisView FORTRAN_NAME(ceedbasisview, CEEDBASISVIEW)
+void fCeedBasisView(CeedInt *basis, FILE *stream, CeedInt *err) {
+//TODO
+//CEED_EXTERN int CeedBasisView(CeedBasis basis, FILE *stream);
+}
+
+#define fCeedBasisApply FORTRAN_NAME(ceedbasisapply, CEEDBASISAPPLY)
+void fCeedBasisApply(CeedInt *basis, CeedInt *tmode, CeedInt *emode, const CeedScalar *u,
+    CeedScalar *v, CeedInt *err) {
+  *err = CeedBasisApply(CeedBasis_dict[*basis], *tmode, *emode, u, v);
+}
+
+#define fCeedBasisGetNumNodes \
+    FORTRAN_NAME(ceedbasisgetnumnodes, CEEDBASISGETNUMNODES)
+void fCeedBasisGetNumNodes(CeedInt *basis, CeedInt *P, CeedInt *err) {
+  *err = CeedBasisGetNumNodes(CeedBasis_dict[*basis], P);
+}
+
+#define fCeedBasisGetNumQuadraturePoints \
+    FORTRAN_NAME(ceedbasisgetnumquadraturepoints, CEEDBASISGETNUMQUADRATUREPOINTS)
+void fCeedBasisGetNumQuadraturePoints(CeedInt *basis, CeedInt *Q, CeedInt *err) {
+  *err = CeedBasisGetNumQuadraturePoints(CeedBasis_dict[*basis], Q);
+}
+
+#define fCeedBasisDestroy FORTRAN_NAME(ceedbasisdestroy,CEEDBASISDESTROY)
+void fCeedBasisDestroy(CeedInt *basis, CeedInt *err) {
+  *err = CeedBasisDestroy(&CeedBasis_dict[*basis]);
+
+  if (*err == 0) {
+    CeedBasis_n--;
+    if (CeedBasis_n == 0) {
+      CeedFree(&CeedBasis_dict);
+      CeedBasis_count = 0;
+      CeedBasis_count_max = 0;
+    }
+  }
+}
+
+#define fCeedGaussQuadrature FORTRAN_NAME(ceedgaussquadrature, CEEDGAUSSQUADRATURE)
+void fCeedGaussQuadrature(CeedInt *Q, CeedScalar *qref1d, CeedScalar *qweight1d,
+    CeedInt *err)
+{
+  *err = CeedGaussQuadrature(*Q, qref1d, qweight1d);
+}
+
+#define fCeedLobattoQuadrature \
+    FORTRAN_NAME(ceedlobattoquadrature, CEEDLOBATTOQUADRATURE)
+void fCeedLobattoQuadrature(CeedInt *Q, CeedScalar *qref1d, CeedScalar *qweight1d,
+    CeedInt *err)
+{
+  *err = CeedLobattoQuadrature(*Q, qref1d, qweight1d);
+}
+
+static CeedQFunction *CeedQFunction_dict = NULL;
+static int CeedQFunction_count = 0;
+static int CeedQFunction_n = 0;
+static int CeedQFunction_count_max = 0;
+
+struct fContext {
+  void (*f)(void *ctx, void *qdata, CeedInt *nq, const CeedScalar *const u,
+           CeedScalar *const v, int *ierr);
+  void *innerctx;
+};
+
+static int CeedQFunctionFortranStub(void *ctx, void *qdata, CeedInt nq,
+    const CeedScalar *const *u, CeedScalar *const *v) {
+  struct fContext *fctx = ctx;
+  int ierr;
+  //Note: Since u and v are arrays to pointers, they
+  //can't be passed as is to Fortran. So, what I do is
+  //find the first non-NULL pointer and pass it is
+  //address to Fortran function. On the Fortran side, to access
+  //specific type of data, the use has to calculate the indices.
```

Comment:
I think the extensible way is to pass a base pointer and an array of offsets. Then you would write `u(offset(f) + i)` to correspond to `u[f][i]` (or `u[f-1][i-1]` if the Fortran indexing starts at 1).

An alternative is to just enumerate calls with explicit arguments, so the Fortran would write something like

    subroutine qf(ctx, qdata, nq, u1, u2, u3, v1, v2, ierr)

and the stub would have to invoke the correct number of arguments (not a big deal to enumerate up to 10 total arguments).